keps/sig-network/2595-expanded-dns-config/README.md

```diff
@@ -0,0 +1,306 @@
+- Expand `MaxDNSSearchListChars` to 2048
+
+### User Stories (Optional)
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+This enhancement relaxes the validation of `Pod` and `PodTemplate`. Once the
+feature is activated, it must be carefully disabled. Otherwise, the objects left
+over from the previous version which have the expanded DNS configuration can be
+problematic.
+
+### Risks and Mitigations
+
+There may be some environments(DNS resolver or others) that break without
+current limitations. At this point, it is fair to open a bug, so they can fix it.
+
+## Design Details
+
+- Declare and define `MaxDNSSearchPathsExpanded` to `32` and
+`MaxDNSSearchListCharsExpanded` to `2048`
+- Add the feature gate `ExpandedDNSConfig` (see [Feature Enablement and
+Rollback](#feature-enablement-and-rollback))
+- If the feature gate `ExpandedDNSConfig` is enabled, replace
+`MaxDNSSearchPaths` with `MaxDNSSearchPathsExpanded` and replace
+`MaxDNSSearchListChars` with `MaxDNSSearchListCharsExpanded` to allow expanded
+DNS configuration
+
+### Test Plan
+
+- Add unit tests of validating expanded DNS config
+
+### Graduation Criteria
+
+#### Alpha
+
+- Implement the feature
+- Add appropriate unit tests
+
+#### Alpha -> Beta Graduation
+
+- Address feedback from alpha
+- Sufficient testing
+
+#### Beta -> GA Graduation
+
+- Address feedback from beta
+- Sufficient number of users using the feature
+- Close on any remaining open issues & bugs
+
+### Upgrade / Downgrade Strategy
+
+N/A
+
+### Version Skew Strategy
+
+In clusters with older kubelets, old kubelets with `resolvConf` configured to
+exceed bounds throw warnings but do not fail. Eventually, old kubelets truncate
+the overage and apply the actual DNS resolver configuration.
+
+In clusters with a replicated control plane, all kube-apiservers should enable
+or disable the expanded DNS configuration feature.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+- **How can this feature be enabled / disabled in a live cluster?**
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: `ExpandedDNSConfig`
+    - Components depending on the feature gate:
+      - `kubelet`
+      - `kube-apiserver`
+  - [ ] Other
+    - Describe the mechanism:
+    - Will enabling / disabling the feature require downtime of the control
+      plane?
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+- **Does enabling the feature change any default behavior?**
+
+Enabling this feature allows kubernetes to have objects(Pod or PodTemplate) with
+the expanded DNS configuration.
+
+- **Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?**
+
+Yes, the feature can be disabled by disabling the feature gate.
+
+Once the feature is disabled, kube-apiserver will reject the pod having expanded
```

Comment:
@gjkim42 what happens to existing pods in the API when the feature gate is disabled, and then something tries to update non-DNS podSpec fields? Do those things just their update denied, even though they didn't change any relevant fields?

Reply:
> what happens to existing pods in the API when the feature gate is disabled

The existing pods will be running with the expanded DNS config.



> and then something tries to update non-DNS podSpec fields? Do those things just their update denied, even though they didn't change any relevant fields?

In v1.22 or higher, you can update non-DNSConfig podSpec fields, because we will add the exception. See [PR](https://github.com/kubernetes/kubernetes/pull/100651).
In v1.21 or lower, any update requests will be denied, if the pod already had expanded DNS config.